core/node/node_framework/src/implementations/layers/web3_api/server.rs

```diff
@@ -0,0 +1,225 @@
+use std::num::NonZeroU32;
+
+use tokio::{sync::oneshot, task::JoinHandle};
+use zksync_core::api_server::web3::{state::InternalApiConfig, ApiBuilder, ApiServer, Namespace};
+
+use crate::{
+    implementations::resources::{
+        healthcheck::AppHealthCheckResource, pools::ReplicaPoolResource,
+        sync_state::SyncStateResource, web3_api::TxSenderResource,
+    },
+    service::{ServiceContext, StopReceiver},
+    task::Task,
+    wiring_layer::{WiringError, WiringLayer},
+};
+
+/// Set of optional variables that can be altered to modify the behavior of API builder.
+#[derive(Debug, Default)]
+pub struct Web3ServerOptionalConfig {
+    pub namespaces: Option<Vec<Namespace>>,
+    pub filters_limit: Option<usize>,
+    pub subscriptions_limit: Option<usize>,
+    pub batch_request_size_limit: Option<usize>,
+    pub response_body_size_limit: Option<usize>,
+    pub websocket_requests_per_minute_limit: Option<NonZeroU32>,
+    pub tree_api_url: Option<String>,
+}
+
+impl Web3ServerOptionalConfig {
+    fn apply(self, mut api_builder: ApiBuilder) -> ApiBuilder {
+        if let Some(namespaces) = self.namespaces {
+            api_builder = api_builder.enable_api_namespaces(namespaces);
+        }
+        if let Some(filters_limit) = self.filters_limit {
+            api_builder = api_builder.with_filter_limit(filters_limit);
+        }
+        if let Some(subscriptions_limit) = self.subscriptions_limit {
+            api_builder = api_builder.with_subscriptions_limit(subscriptions_limit);
+        }
+        if let Some(batch_request_size_limit) = self.batch_request_size_limit {
+            api_builder = api_builder.with_batch_request_size_limit(batch_request_size_limit);
+        }
+        if let Some(response_body_size_limit) = self.response_body_size_limit {
+            api_builder = api_builder.with_response_body_size_limit(response_body_size_limit);
+        }
+        if let Some(websocket_requests_per_minute_limit) = self.websocket_requests_per_minute_limit
+        {
+            api_builder = api_builder
+                .with_websocket_requests_per_minute_limit(websocket_requests_per_minute_limit);
+        }
+        api_builder = api_builder.with_tree_api(self.tree_api_url);
+        api_builder
+    }
+}
+
+/// Internal-only marker of chosen transport.
+#[derive(Debug, Clone, Copy)]
+enum Transport {
+    Http,
+    Ws,
+}
+
+pub struct Web3ServerLayer {
+    transport: Transport,
+    port: u16,
+    internal_api_config: InternalApiConfig,
+    optional_config: Web3ServerOptionalConfig,
+}
+
+impl std::fmt::Debug for Web3ServerLayer {
```

Comment:
Dumb question: What's the problem with deriving `Debug` here?

Reply:
Leftover from WIP state. I was experimenting and had some tries to expose the builder that we already have for API, and at that time `Debug` could've not been derived. Then I gave up and went with an extra type for builder fields and forgot to change the implementation.

Thanks for noticing!